content/en/blog/2023-01-30-building-a-raft-part-3.md

```diff
@@ -0,0 +1,110 @@
+
+!["State change flowchart"](/img/blog/2023-01-19-building-a-raft-part-3/stateChanges.png)
+
+If a Raft server does not receive any communication from the leader (heartbeat or regular requests) for a given amount of time, it will assume the leader has died and will start a campaign to become the new leader, incrementing it’s term, voting for itself and sending `RequestVote` RPCs to the other servers in the cluster. If it receives a majority of the cluster’s votes, it declares itself leader and starts sending out heartbeats. Luckily for us, the `RequestVote` RPC ends up being much simpler than `AppendEntries`, but we should still spend some time stepping through it.
+
+## RequestVote Steps
+
+The `RequestVote` function is how candidates carry out elections in remote replicas. The function expects as input a `VoteRequest`, which contains the candidate's term, it's unique identifier, and some details of it's last log entry.
+
+*Note: Abbreviated for brevity*
+```proto
+message VoteRequest {
+    int32 term = 2;           // The candidate's term.
+    string candidateId = 3;   // ID of candidate requesting a vote.
+    int64 lastLogIndex = 4;   // The index of candidate’s last log entry
+    int32 lastLogTerm = 5;    // The term number of candidate’s last log entry
+}
+```
+
+In return, each voter will respond with a `VoteReply`, which contains their own unique identifier, their perspective of the term, and their vote.
+
+```proto
+message VoteReply {
+    string id = 1;            // ID of the Raft node sending the vote.
+    int32 term = 2;           // Current term number
+    bool voteGranted = 3;     // True if the voter has voted for the candidate
+}
+```
+
+#### 1. Vote "no" if the candidate is behind
+
+Just like in `AppendEntries`, we first need to check if the candidate has an up-to-date term. If a candidate requesting a vote has a term less than the recipient's term, that candidate is not up-to-date with the rest of the cluster, and should be denied the vote.
+
+#### 2. Vote "yes" if the candidate is ahead or even (but only vote once!)
+
+Each server in a Raft cluster keeps track of who it has most recently voted for (until the next successful heartbeat). If this `votedFor` field is null, then it hasn’t voted for anyone in the current term and is free to vote for the candidate that sent the request.
+
+Here's the full logic for the `RequestVote` function:
+
+```golang
+func (s *RaftServer) RequestVote(ctx context.Context, in *api.VoteRequest) (out *api.VoteReply, err error) {
+	s.Lock()
+	defer s.Unlock()
+	out = &api.VoteReply{Id: s.id, Term: s.currentTerm}
+	lastLogIndex, lastLogTerm := s.lastLogIndexAndTerm()
+
+	if in.Term > s.currentTerm {
+		fmt.Printf("RequestVote: in.Term > s.currentTerm, reverting to follower\n")
+		s.becomeFollower(in.Term)
+	}
```

Comment:
@DanielSollis can you please add a subsection of text to the enumerated steps above to explain what is happening in these 4 lines?